Add quirks for CTM Lyng outlets/relays and cooktop guard

uvNikita opened the pull request:
## Proposed change
CTM Lyng is a norwegian manufacturer which is very nice to work with.
They provided me with zigbee spec docs for their Tavrele 25A product.

The product mostly works out of the box, but there is a couple of manufacturer specific clusters (client `CTMDiagnosticsCluster` and server `CTMStoveGuardCluster`) that could be useful to have.

Once I'll get their stove guard sensor, I'll see how it can also be integrated.


## Additional information
[Outlet Zigbee Specs](https://github.com/user-attachments/files/20724763/ctm_mains_power_outlet_1.0.docx)
[Stove Guard Zigbee Specs](https://github.com/user-attachments/files/20758273/ctm_mkomfy_1.0.docx)



## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

uvNikita committed:
Add quirks for CTM Lyng outlets and cooktop guards

uvNikita committed:
Add quirks for CTM Lyng outlets and cooktop guards

puddly commented on zhaquirks/ctm/__init__.py:
```diff
@@ -0,0 +1,264 @@
+
+        ctm_alarm_status: Final = foundation.ZCLAttributeDef(
+            id=0x0001,
+            type=AlarmStatus,
```
For these devices, set `zcl_type=foundation.DataTypeId.uint8` and then you can have `AlarmStatus` be an `enum8` subclass.

For the vast majority of devices the ZCL type can be inferred from the Python type but there are many devices that do not follow the specification and use bare integers instead of enums.

uvNikita commented:
@puddly I set `zcl_type`  on enum attributes as you said.
I also added sever and client commands and a stoveguard support (not tested fully until I receive it, but should work in principle).

I have a couple of questions:

- For stoveguard I might also need to setup a Bus. Is there a way to do it in quirks v2?
- As you can see, their outlet has CTMStoveGuardCluster as a client cluster and stoveguard has it as a server cluster. Should I split them into two separate classes so that all attributes/commands are not present in both devices?
- I tried to add `paired_with` attribute from the client CTMStoveGuardCluster as an entity, but zha fails with `AssertionError` pointing to [assert cluster_handler](https://github.com/zigpy/zha/blob/a0fd5acf6b7676a147b390fa5eb7ac6a0e9e4a23/zha/application/discovery.py#L237). Reading the attribute manually from "Manage Zigbee device" works correctly.

uvNikita committed:
Rename stoveguard -> cooktop guard

uvNikita committed:
Rename stoveguard -> cooktop guard

uvNikita commented:
@TheJulianJES Could you look at this one when you have time?

I finally installed cooktop guard from CTM Lyng that pairs with their relay/outlet. Renamed stoveguard to cooktop guard.

I still have a few issues / questions:

- ~~Is there a way to setup bus in quirks v2?~~
- Should I split CTMCooktopGuardCluster into two CTMCooktopGuardServerCluster and CTMCooktopGuardClientCluster?
- `paired_with` / `cooktop_guard_sensor_address` entity (see commented lines) still fails with `AssertionError` pointing to [assert cluster_handler](https://github.com/zigpy/zha/blob/a0fd5acf6b7676a147b390fa5eb7ac6a0e9e4a23/zha/application/discovery.py#L237). It's probably because it's a client Cluster, but not sure how to fix this.
- ~~`alarm_status`, `battery_status` and `active_status` shows up and works correctly, but only if I update them manually. They don't seem to update automatically no matter what I do. AFAIK, this sensor "sleeps" until paired relay reports electrical usage to preserve battery. After that it wakes up and suppose to report temperature/status updates. I can confirm that manual reading works after there is electrical load, but no automatic updates. Maybe there is an obvious fix for that?~~

uvNikita committed:
CTM Lyng cooktop guard fixes

- Fix reporting config for ctm_alarm_status, ctm_battery_alarm, ctm_cooktop_active by specifying correct type
- Replace BatteryStatus, ActiveStatus with binary sensors

uvNikita committed:
CTM Lyng cooktop guard fixes

- Fix reporting config for ctm_alarm_status, ctm_battery_alarm, ctm_cooktop_active by specifying correct type
- Replace BatteryStatus, ActiveStatus with binary sensors

uvNikita commented:
I managed to fix reporting configuration for `ctm_alarm_status`, `ctm_battery_alarm` and `ctm_cooktop_active` attributes.

@puddly If I used `enum8` in `type=<enum>`, and `zcl_type=foundation.DataTypeId.uint8`, then I was getting the following error on reporting configuration (`INVALID_DATA_TYPE`):


```
2025-09-02 12:53:17.613 DEBUG (MainThread) [zigpy.zcl] [0x7703:1:0xffc9] Decoded ZCL frame: CtmMKomfyCluster:Configure_Reporting_rsp(status_records=[ConfigureReportingResponseRecord(status=<Status.INVALID_DATA_TYPE: 141>, direction=<ReportingDirection.SendReports: 0>, attrid=1), ConfigureReportingResponseRecord(status=<Status.INVALID_DATA_TYPE: 141>, direction=<ReportingDirection.SendReports: 0>, attrid=5)])
```

Changing type back to `t.uint8_t` solved the problem.

At this point everything is working pretty good, I only have a couple of questions left:

- Should I split CTMCooktopGuardCluster into two CTMCooktopGuardServerCluster and CTMCooktopGuardClientCluster?
- `paired_with` / `cooktop_guard_sensor_address` entity in the outlet device (see commented lines) still fails with `AssertionError` pointing to [assert cluster_handler](https://github.com/zigpy/zha/blob/a0fd5acf6b7676a147b390fa5eb7ac6a0e9e4a23/zha/application/discovery.py#L237). It's probably because it's a client Cluster, but not sure how to fix this.

TheJulianJES commented:
I'll get back to you on the rest, but the `zcl_type` causing issues with attribute reporting should be fixed with:
- https://github.com/zigpy/zigpy/pull/1650

uvNikita committed:
Set zcl_type for ctm_alarm_status

uvNikita committed:
Set zcl_type for ctm_alarm_status

uvNikita commented:
> I'll get back to you on the rest, but the `zcl_type` causing issues with attribute reporting should be fixed with:
> 
>     * [Use attribute zcl_type when configuring reporting. zigpy#1650](https://github.com/zigpy/zigpy/pull/1650)

That's great, then I changed it back to setting `zcl_type=foundation.DataTypeId.uint8` since it should be fixed in the recent version of zigpy.

uvNikita committed:
ctm: use manufacturer_code, fix command directions and device class

- Replace is_manufacturer_specific=True with manufacturer_code=0x1337 on
  attribute definitions (the old approach now emits deprecation warnings
  in zigpy).
- Use BaseAttributeDefs/BaseCommandDefs for the fully custom CTM
  diagnostics and cooktop guard clusters.
- Swap ServerCommandDefs and ClientCommandDefs in CTMCooktopGuardCluster
  to match the cluster spec: pair_with_sensor/relay_status are received
  by the server, while pair_with_relay/relay_status_request/on_command/
  alarm_command are generated by the server.
- Drop the redundant zcl_type override on ctm_alarm_status (derived from
  the enum8 type).
- Add ep_attribute to CTMDiagnosticsCluster.
- Fix cooktop temperature sensor to use SensorDeviceClass.TEMPERATURE
  instead of NumberDeviceClass.TEMPERATURE.

uvNikita committed:
ctm: use manufacturer_code, fix command directions and device class

- Replace is_manufacturer_specific=True with manufacturer_code=0x1337 on
  attribute definitions (the old approach now emits deprecation warnings
  in zigpy).
- Use BaseAttributeDefs/BaseCommandDefs for the fully custom CTM
  diagnostics and cooktop guard clusters.
- Swap ServerCommandDefs and ClientCommandDefs in CTMCooktopGuardCluster
  to match the cluster spec: pair_with_sensor/relay_status are received
  by the server, while pair_with_relay/relay_status_request/on_command/
  alarm_command are generated by the server.
- Drop the redundant zcl_type override on ctm_alarm_status (derived from
  the enum8 type).
- Add ep_attribute to CTMDiagnosticsCluster.
- Fix cooktop temperature sensor to use SensorDeviceClass.TEMPERATURE
  instead of NumberDeviceClass.TEMPERATURE.

uvNikita committed:
ctm: remove commented-out client cluster sensor

uvNikita committed:
ctm: remove commented-out client cluster sensor

uvNikita committed:
ctm: add device classes to cooktop guard binary sensors

Set BinarySensorDeviceClass.BATTERY on the battery alarm and
BinarySensorDeviceClass.RUNNING on the cooktop active binary sensor,
and coerce their uint8 values to bool so HA renders clean on/off
states.

uvNikita committed:
ctm: add device classes to cooktop guard binary sensors

Set BinarySensorDeviceClass.BATTERY on the battery alarm and
BinarySensorDeviceClass.RUNNING on the cooktop active binary sensor,
and coerce their uint8 values to bool so HA renders clean on/off
states.

TheJulianJES commented on zhaquirks/ctm/__init__.py:
```diff
@@ -0,0 +1,308 @@
+        )
+        ctm_ambient_temperature: Final = ZCLAttributeDef(
+            id=0x0004,
+            type=t.uint8_t,
```
Z2M has this as `t.uint16_t`, not sure what is correct. Debug logs should include this when a report/read comes in.

uvNikita replied:
Here is log when I enabled debug logging and read the value of this attribute:

`DEBUG (MainThread) [zigpy.zcl] [0x7703:1:0xffc9] Decoded ZCL frame: CTMCooktopGuardCluster:ReadAttributesResponse(status_records=[ReadAttributeRecord(attrid=4, status=<Status.SUCCESS: 0>, value=TypeValue(type=uint8_t, value=28))])`

Looks like uint8_t is correct? And this is also what's specified in the spec docs in the first post.

TheJulianJES reviewed (commented):
Looks pretty good! Very small things:

https://github.com/zigpy/zha-device-handlers/pull/4126/changes/5c035f5bb890e4155ef086af675409d7625f15b5 dropped the `zcl_type` for an attribute that's seemingly an int on the ZCL side, but an enum on the zigpy side. I think it was correct before and it's broken now? Though it's really only an issue if you were to try and write to that attribute. For reading, zigpy should correctly parse it either way.

Also, the usual order for kwargs in quirks v2 entities is this:
- attribute_name, cluster_id, endpoint_id first; translation_key, fallback_name last.

Other feedback: 
- nit: name casing: "CTMDiagnostics" vs "CtmCooktopGuard"

uvNikita committed:
ctm: restore zcl_type=uint8 for AlarmStatus attributes

uvNikita committed:
ctm: reorder v2 entity kwargs to conventional style

uvNikita committed:
ctm: fix cluster name casing for consistency

uvNikita commented:
@TheJulianJES Thank you for the review, made all the requested changes.

uvNikita commented:
I rebased this PR on top of the current dev branch.

@TheJulianJES I think this one is ready for merge?

zigpy-review-bot committed:
Update imports for the quirks API move (#5113)

uvNikita committed:
ctm: address PR review feedback

- Use ZCL_CLUSTER_REVISION_ATTR for cluster_revision instead of a
  manufacturer-specific 0xFFFD attribute
- Set manufacturer_code on cooktop guard commands instead of
  is_manufacturer_specific=True
- Declare reported cooktop guard attributes as access="rp"
- Convert relay address (EUI64) sensor value to str for HA
- Add state_class/device_class to temperature, counter and duration
  sensors

uvNikita commented:
- Changed to ZCL_CLUSTER_REVISION_ATTR
- Use manufacturer_code in commands
- Set access="rp"
- Convert EUI64 value to str
- Add state_class/device_class to sensors
- Rebased on top of the current dev.
 
Diagnostics files:
- [CTM Lyng mKomfy.json](https://github.com/user-attachments/files/30581734/zha-e6a6134e87a1487b579906f1f839c1fd-CTM.Lyng.mKomfy-9f09588ee4699d8489728d5cede38fb9.json)
- [CTM Lyng Tavlerele 25A.json](https://github.com/user-attachments/files/30581738/zha-e6a6134e87a1487b579906f1f839c1fd-CTM.Lyng.Tavlerele.25A-3ebeb58a02361c74c3ac1196a13d4001.json)

TheJulianJES commented on zhaquirks/ctm/outlet.py:
```diff
@@ -0,0 +1,59 @@
+        entity_type=EntityType.DIAGNOSTIC,
+        initially_disabled=True,
+        translation_key="reboot_counter",
+        fallback_name="Reboot counter",
```
IMO "Reboot count" sounds better than both "Reboot counter" and "Rebooted count", but [we already have translations](https://github.com/home-assistant/core/blob/7289e79c4b860a783145d6d35a03dc2654861207/homeassistant/components/zha/strings.json#L1823-L1825) for `rebooted_count`: "Rebooted count", so let's use that:

```suggestion
        translation_key="rebooted_count",
        fallback_name="Rebooted count",
```

TheJulianJES commented on zhaquirks/ctm/outlet.py:
```diff
@@ -0,0 +1,59 @@
+        entity_type=EntityType.DIAGNOSTIC,
+        initially_disabled=True,
+        translation_key="button_click_counter",
+        fallback_name="Button click counter",
```
Similar to above, let's use "count" everywhere, instead of "counter":
```suggestion
        translation_key="button_click_count",
        fallback_name="Button click count",
```

TheJulianJES commented on zhaquirks/ctm/cooktop_guard.py:
```diff
@@ -0,0 +1,107 @@
+        entity_type=EntityType.DIAGNOSTIC,
+        initially_disabled=True,
+        translation_key="reboot_counter",
+        fallback_name="Reboot counter",
```
See above comments:
```suggestion
        translation_key="rebooted_count",
        fallback_name="Rebooted count",
```

TheJulianJES reviewed (commented):
Two small things but I think this looks good otherwise. Thanks and sorry for the wait!
The reporting config mentioned by the bot might be worth addressing as well. I think `max_interval=3600`, so the device reports the attribute value at least every hour, might be nice.

uvNikita committed:
ctm: rename count entities and add reporting refresh interval

- Rename ctm_reboot_counter -> ctm_rebooted_count and use rebooted_count
  translation_key (matches existing HA translation)
- Rename ctm_button_0_click_counter -> ctm_button_0_click_count and use
  'count' naming in entity names
- Set max_interval=3600 on cooktop guard reporting configs to refresh
  stale values while keeping min_interval=0 for immediate alarm delivery

uvNikita commented:
Done. I also renamed the attributes for consistency, or did you want me to only change translation_key/fallback_name?
